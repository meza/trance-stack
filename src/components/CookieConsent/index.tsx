import React from 'react';
import { Form } from '@remix-run/react';

export interface ConsentData {
  analytics?: boolean | undefined;
  marketing?: boolean | undefined;
  performance?: boolean | undefined;
}

interface CookieConsentContextProps {
  analytics?: boolean | undefined;
  marketing?: boolean | undefined;
  performance?: boolean | undefined;
  enableAnalytics: () => void;
  enableMarketing: () => void;
  enablePerformance: () => void;
}

export const CookieConsentContext = React.createContext<CookieConsentContextProps>({} as CookieConsentContextProps);

export const CookieConsentProvider = ({ children, consentData }: { children: React.ReactNode, consentData?: ConsentData | undefined }) => {
  const [analytics, setAnalytics] = React.useState(consentData?.analytics);
  const [marketing, setMarketing] = React.useState(consentData?.marketing);
  const [performance, setPerformance] = React.useState(consentData?.performance);
  const enableAnalytics = () => {
    setAnalytics(true);
  };

  const enableMarketing = () => {
    setMarketing(true);
  };

  const enablePerformance = () => {
    setPerformance(true);
  };

  return (
    <CookieConsentContext.Provider
      value={{
        analytics: analytics,
        marketing: marketing,
        performance: performance,
        enableAnalytics: enableAnalytics,
        enableMarketing: enableMarketing,
        enablePerformance: enablePerformance
      }}>
      {children}
    </CookieConsentContext.Provider>
  );
};

export const CookieConsentBanner = () => {
  const { analytics, marketing, performance } = React.useContext(CookieConsentContext);
  const [analyticsConsent, setAnalyticsConsent] = React.useState(analytics || false);
  const [marketingConsent, setMarketingConsent] = React.useState(marketing || false);
  const [performanceConsent, setPerformanceConsent] = React.useState(performance || false);

  if (analytics !== undefined && marketing !== undefined && performance !== undefined) {
    return null;
  }

  return (
    <div className="cookie-consent-banner">
      <Form method="post" action="/settings/cookie-consent" replace reloadDocument={true}>
        <label htmlFor={'analytics'}>Analytics</label>
        <input type={'checkbox'} name={'analytics'} id={'analytics'} defaultChecked={true} onClick={() => setAnalyticsConsent(!analyticsConsent)}/>
        <label htmlFor={'marketing'}>Marketing</label>
        <input type={'checkbox'} name={'marketing'} id={'marketing'} defaultChecked={marketingConsent} onClick={() => setMarketingConsent(!marketingConsent)}/>
        <label htmlFor={'performance'}>Performance</label>
        <input type={'checkbox'} name={'performance'} id={'performance'} defaultChecked={true} onClick={() => setPerformanceConsent(!performanceConsent)}/>
        <button type="submit">Save</button>
      </Form>
    </div>
  );
};
