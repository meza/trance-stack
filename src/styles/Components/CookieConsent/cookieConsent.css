.cookie-consent-container {
  background-color: var(--background-2);
  bottom: 0;
  left: 0;
  padding: 2rem;
  position: absolute;
  text-align: justify;
  min-width: 30ch;
  max-width: 50ch;
  width: 80dvw;

  @media (--motion-ok) {
    animation: slide-in-bottom 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
  }

  & .icon-cookie {
    height: 2.2rem;
    width: 2.2rem;
  }

  & .cookie-consent-header {
    display: grid;
    grid-auto-flow: column;
    justify-content: left;
    align-items: baseline;
    font-size: 1.5rem;
  }

  & .cookie-consent-text {
    padding: 1rem;
  }

  & .cookie-consent-switches {
    grid-auto-flow: column;
    display: grid;
    grid-column-gap: 2rem;
    justify-content: center;
    margin-bottom: 1.5rem;
    margin-top: 1rem;
    padding-left: 1.5rem;
    padding-right: 1.5rem;

    & .cookie-consent-switch {
      display: grid;
      grid-auto-flow: row;
      justify-self: center;
    }
  }

  & .cookie-consent-buttons {
    display: grid;
    grid-auto-flow: column;
    justify-content: space-between;
  }

}
