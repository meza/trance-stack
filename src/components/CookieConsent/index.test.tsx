import React, { useContext, useEffect } from 'react';
import { redirect } from '@remix-run/node';
import { unstable_createRemixStub as createRemixStub } from '@remix-run/testing';
import { renderWithi18n } from '@test';
import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi, it, describe } from 'vitest';
import { CookieConsentBanner, CookieConsentContext, CookieConsentProvider } from '~/components/CookieConsent/index';

const TestComponent = () => {
  const { analytics } = useContext(CookieConsentContext);
  return <>{analytics === undefined ? 'undefined' : String(analytics)}</>;
};

vi.mock('~/atoms/Icons/cookie', () => ({
  default: () => <div id={'cookie-icon'}/>
}));

describe('The Cookie Consent Component', () => {
  describe('the provider', () => {
    it('does not interfere with the unset status', async () => {
      render(
        <CookieConsentProvider>
          <TestComponent/>
        </CookieConsentProvider>
      );

      await waitFor(() => expect(screen.getByText('undefined')).toBeInTheDocument());
    });

    it('can pass on the initial consent data', async () => {
      render(
        <CookieConsentProvider
          consentData={
            {
              analytics: false
            }
          }>
          <TestComponent/>
        </CookieConsentProvider>
      );

      await waitFor(() => expect(screen.getByText('false')).toBeInTheDocument());
    });

    it('propagates the change', async () => {
      const TestComponent1 = () => {
        const { setAnalytics } = useContext(CookieConsentContext);
        useEffect(() => {
          setAnalytics(false);
        });
        return <></>;
      };

      render(
        <CookieConsentProvider>
          <TestComponent/>
          <TestComponent1/>
        </CookieConsentProvider>
      );

      await waitFor(() => expect(screen.getByText('false')).toBeInTheDocument());
    });
  });

  describe('the banner', () => {
    it('renders the form', async ({ expect }) => {
      const RemixStub = createRemixStub([
        {
          path: '/*',
          element: <CookieConsentProvider><CookieConsentBanner/></CookieConsentProvider>
        }
      ]);

      const component = renderWithi18n(<RemixStub/>);
      expect(component.asFragment()).toMatchSnapshot();
    });

    it('hides and saves the form correctly on deny', async ({ expect }) => {
      let analytics, marketing;
      const settingsSpy = vi.fn();
      const user = userEvent.setup();
      const RemixStub = createRemixStub([
        {
          path: '/',
          element: <CookieConsentProvider><CookieConsentBanner/></CookieConsentProvider>
        },
        {
          path: '/settings/cookie-consent',
          action: async ({ request }) => {
            const formData = await request.formData();
            analytics = formData.get('analytics');
            marketing = formData.get('marketing');
            settingsSpy();
            throw redirect('/');
          }
        }
      ]);

      renderWithi18n(<RemixStub initialEntries={['/']}/>);
      const denyButton = screen.getByRole('button', { name: 'cookieConsent.deny' });
      const dialog = screen.getByRole('dialog');
      expect(dialog).toBeInTheDocument();

      await user.click(denyButton);

      expect(settingsSpy).toHaveBeenCalled();
      expect(analytics).toBe('false');
      expect(marketing).toBe('false');
      expect(dialog).not.toBeInTheDocument();
    });

    it.each([undefined, true, false])('is wired up correctly when %s', async (analyticsState) => {
      // As long as the form is wired up properly, we trust that the browser can do its thing.
      // We don't need to test that the submission works because there is no custom logic in the
      // submission handler.
      const RemixStub = createRemixStub([
        {
          path: '/',
          element: <CookieConsentProvider
            consentData={
              {
                analytics: analyticsState
              }
            }
          ><CookieConsentBanner/></CookieConsentProvider>
        }
      ]);

      renderWithi18n(<RemixStub initialEntries={['/']}/>);
      const dialog = screen.getByRole('dialog');
      expect(dialog).toMatchSnapshot();
    });

    it('adds the escape key handler', async ({ expect }) => {
      const keydownSpy = vi.fn();
      const user = userEvent.setup();
      const RemixStub = createRemixStub([
        {
          path: '/',
          element: <CookieConsentProvider><CookieConsentBanner/></CookieConsentProvider>
        }
      ]);

      renderWithi18n(<RemixStub initialEntries={['/']}/>);
      const dialog = screen.getByRole('dialog');

      // This is rather ugly but since the form submit isn't implemented in jsdom, it kinda evens out
      // and gets the job done.
      dialog.querySelector('form')!.requestSubmit = () => {
        keydownSpy();
      };

      expect(dialog).toBeInTheDocument();

      await user.keyboard('[Escape]');

      expect(keydownSpy).toHaveBeenCalled();
    });
  });
});
