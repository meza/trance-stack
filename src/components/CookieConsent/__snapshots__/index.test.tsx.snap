// Vitest Snapshot v1

exports[`The Cookie Consent Component > the banner > is wired up correctly 1`] = `
<div
  aria-describedby="cookie-consent-text"
  aria-labelledby="cookie-consent-header"
  class="cookie-consent-container"
  role="dialog"
>
  <div
    class="cookie-consent-header"
    id="cookie-consent-header"
  >
    <div
      id="cookie-icon"
    />
    cookieConsent.title
  </div>
  <div
    class="cookie-consent-text"
    id="cookie-consent-text"
  >
    cookieConsent.disclaimer
  </div>
  <form
    action="/settings/cookie-consent"
    id="cookie-consent-form-accept"
    method="post"
  >
    <div
      class="cookie-consent-switches"
    >
      <div
        class="cookie-consent-switch"
      >
        <label
          class="cookie-consent-title"
          for="cookie-consent-necessary"
        >
          cookieConsent.label.necessary
        </label>
        <label
          class="toggle-component cookie-consent-switch"
        >
          <input
            aria-hidden="true"
            name="necessary"
            type="hidden"
            value="true"
          />
          <input
            checked=""
            disabled=""
            id="cookie-consent-necessary"
            role="switch"
            tabindex="0"
            type="checkbox"
          />
          <span
            aria-hidden="true"
            class="slider"
          />
        </label>
      </div>
      <div
        class="cookie-consent-switch"
      >
        <label
          class="cookie-consent-title"
          for="cookie-consent-analytics"
        >
          cookieConsent.label.analytics
        </label>
        <label
          class="toggle-component cookie-consent-switch"
        >
          <input
            aria-hidden="true"
            name="analytics"
            type="hidden"
            value="true"
          />
          <input
            checked=""
            id="cookie-consent-analytics"
            role="switch"
            tabindex="1"
            type="checkbox"
          />
          <span
            aria-hidden="true"
            class="slider"
          />
        </label>
      </div>
      <div
        class="cookie-consent-switch"
      >
        <label
          class="cookie-consent-title"
          for="cookie-consent-marketing"
        >
          cookieConsent.label.marketing
        </label>
        <label
          class="toggle-component cookie-consent-switch"
        >
          <input
            aria-hidden="true"
            name="marketing"
            type="hidden"
            value="false"
          />
          <input
            disabled=""
            id="cookie-consent-marketing"
            role="switch"
            tabindex="0"
            type="checkbox"
          />
          <span
            aria-hidden="true"
            class="slider"
          />
        </label>
      </div>
    </div>
  </form>
  <form
    action="/settings/cookie-consent"
    id="cookie-consent-form-deny"
    method="post"
  >
    <input
      name="analytics"
      type="hidden"
      value="false"
    />
    <input
      name="marketing"
      type="hidden"
      value="false"
    />
  </form>
  <div
    class="cookie-consent-buttons"
  >
    <button
      form="cookie-consent-form-deny"
      tabindex="3"
      type="submit"
    >
      cookieConsent.deny
    </button>
    <button
      class="primary"
      form="cookie-consent-form-accept"
      tabindex="2"
      type="submit"
    >
      cookieConsent.accept
    </button>
  </div>
</div>
`;

exports[`The Cookie Consent Component > the banner > renders the form 1`] = `
<DocumentFragment>
  <div
    aria-describedby="cookie-consent-text"
    aria-labelledby="cookie-consent-header"
    class="cookie-consent-container"
    role="dialog"
  >
    <div
      class="cookie-consent-header"
      id="cookie-consent-header"
    >
      <div
        id="cookie-icon"
      />
      cookieConsent.title
    </div>
    <div
      class="cookie-consent-text"
      id="cookie-consent-text"
    >
      cookieConsent.disclaimer
    </div>
    <form
      action="/settings/cookie-consent"
      id="cookie-consent-form-accept"
      method="post"
    >
      <div
        class="cookie-consent-switches"
      >
        <div
          class="cookie-consent-switch"
        >
          <label
            class="cookie-consent-title"
            for="cookie-consent-necessary"
          >
            cookieConsent.label.necessary
          </label>
          <label
            class="toggle-component cookie-consent-switch"
          >
            <input
              aria-hidden="true"
              name="necessary"
              type="hidden"
              value="true"
            />
            <input
              checked=""
              disabled=""
              id="cookie-consent-necessary"
              role="switch"
              tabindex="0"
              type="checkbox"
            />
            <span
              aria-hidden="true"
              class="slider"
            />
          </label>
        </div>
        <div
          class="cookie-consent-switch"
        >
          <label
            class="cookie-consent-title"
            for="cookie-consent-analytics"
          >
            cookieConsent.label.analytics
          </label>
          <label
            class="toggle-component cookie-consent-switch"
          >
            <input
              aria-hidden="true"
              name="analytics"
              type="hidden"
              value="true"
            />
            <input
              checked=""
              id="cookie-consent-analytics"
              role="switch"
              tabindex="1"
              type="checkbox"
            />
            <span
              aria-hidden="true"
              class="slider"
            />
          </label>
        </div>
        <div
          class="cookie-consent-switch"
        >
          <label
            class="cookie-consent-title"
            for="cookie-consent-marketing"
          >
            cookieConsent.label.marketing
          </label>
          <label
            class="toggle-component cookie-consent-switch"
          >
            <input
              aria-hidden="true"
              name="marketing"
              type="hidden"
              value="false"
            />
            <input
              disabled=""
              id="cookie-consent-marketing"
              role="switch"
              tabindex="0"
              type="checkbox"
            />
            <span
              aria-hidden="true"
              class="slider"
            />
          </label>
        </div>
      </div>
    </form>
    <form
      action="/settings/cookie-consent"
      id="cookie-consent-form-deny"
      method="post"
    >
      <input
        name="analytics"
        type="hidden"
        value="false"
      />
      <input
        name="marketing"
        type="hidden"
        value="false"
      />
    </form>
    <div
      class="cookie-consent-buttons"
    >
      <button
        form="cookie-consent-form-deny"
        tabindex="3"
        type="submit"
      >
        cookieConsent.deny
      </button>
      <button
        class="primary"
        form="cookie-consent-form-accept"
        tabindex="2"
        type="submit"
      >
        cookieConsent.accept
      </button>
    </div>
  </div>
</DocumentFragment>
`;
